File: app/src/main/java/com/przemyslawlusnia/vocabularycreator/wordlist/presentation/WordsAdapter.java
package com.przemyslawlusnia.vocabularycreator.wordlist.presentation;

import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.ViewGroup;
import com.przemyslawlusnia.vocabularycreator.core.Constants;
import com.przemyslawlusnia.vocabularycreator.core.utils.ObjectUtils;
import com.przemyslawlusnia.vocabularycreator.wordlist.presentation.viewholder.LearnedWordRecyclerViewHolder;
import com.przemyslawlusnia.vocabularycreator.wordlist.presentation.viewholder.TrainingWordRecyclerViewHolder;
import com.przemyslawlusnia.vocabularycreator.wordlist.presentation.viewholder.WordsViewHolder;
import java.util.ArrayList;
import java.util.List;

public class WordsAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

  private static final String TAG = WordsAdapter.class.getSimpleName();
  private final OnWordsSelectionListener listener;
  private List<WordViewModel> words;
  private int selectedItemsCount;

  public WordsAdapter(OnWordsSelectionListener wordsView) {
    words = new ArrayList<>();
    this.listener = wordsView;
  }

  @Override
  public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
    if (viewType == Constants.TYPE_TRAINING) {
      return new TrainingWordRecyclerViewHolder(parent);
    } else if (viewType == Constants.TYPE_LEARNED) {
      return new LearnedWordRecyclerViewHolder(parent);
    }
    Log.e(TAG, "Unknown Word type");
    return new TrainingWordRecyclerViewHolder(parent);
  }

  @Override
  public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
    final WordViewModel word = words.get(position);
    ((WordsViewHolder) holder).bind(word);
    holder.itemView.setSelected(word.isSelected());
    holder.itemView.setOnClickListener((view) -> {
      boolean newSelection = !word.isSelected();
      word.setSelected(newSelection);
      view.setSelected(newSelection);
      updateSelection(newSelection ? selectedItemsCount + 1 : selectedItemsCount - 1);
    });
  }

  private void updateSelection(int newSelectedItemsCount) {
    selectedItemsCount = newSelectedItemsCount;

    if (ObjectUtils.isNotNull(TAG, "updateSelection", listener)) {
      if (selectedItemsCount > 1) {
        listener.updateMultipleSelection();
      } else if (selectedItemsCount == 1) {
        listener.updateSingleSelection();
      } else {
        listener.updateNoSelection();
      }
    }
  }

  @Override
  public int getItemViewType(int position) {
    WordViewModel word = words.get(position);
    return word.getType();
  }

  @Override
  public int getItemCount() {
    return words.size();
  }

  public void setWords(List<WordViewModel> words) {
    this.words = words;
    notifyDataSetChanged();
  }

  public void addWord(WordViewModel word) {
    words.add(word);
    notifyDataSetChanged();
  }

  public void clearSelection() {
    updateSelection(0);
  }

  public WordViewModel getSelectedWord() {
    int selectedIndex = getFirstSelectedIndex();
    return selectedIndex >= 0 ? words.get(selectedIndex) : new WordViewModel();
  }

  public List<WordViewModel> getSelectedWords() {
    List<WordViewModel> result = new ArrayList<>();
    for (WordViewModel word : words) {
      if (word.isSelected()) {
        result.add(word);
      }
    }
    return result;
  }

  private int getFirstSelectedIndex() {
    for (int i = 0; i < words.size(); i++) {
      if (words.get(i).isSelected()) {
        return i;
      }
    }
    return -1;
  }

  public void editSelectedWord(WordViewModel word) {
    words.set(getFirstSelectedIndex(), word);
    notifyDataSetChanged();
  }
}

File: app/src/main/java/com/przemyslawlusnia/vocabularycreator/wordlist/presentation/WordsAdapter.kt
package com.przemyslawlusnia.vocabularycreator.wordlist.presentation

import android.support.v7.widget.RecyclerView
import android.util.Log
import android.view.ViewGroup
import com.przemyslawlusnia.vocabularycreator.core.Constants
import com.przemyslawlusnia.vocabularycreator.core.utils.ObjectUtils
import com.przemyslawlusnia.vocabularycreator.wordlist.presentation.viewholder.LearnedWordRecyclerViewHolder
import com.przemyslawlusnia.vocabularycreator.wordlist.presentation.viewholder.TrainingWordRecyclerViewHolder
import com.przemyslawlusnia.vocabularycreator.wordlist.presentation.viewholder.WordsViewHolder
import java.util.*

class WordsAdapter(private val listener: OnWordsSelectionListener) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private val TAG = WordsAdapter::class.java.simpleName

    private var words: MutableList<WordViewModel> = ArrayList()
    private var selectedItemsCount: Int = 0

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        if (viewType == Constants.TYPE_TRAINING) {
            return TrainingWordRecyclerViewHolder(parent)
        } else if (viewType == Constants.TYPE_LEARNED) {
            return LearnedWordRecyclerViewHolder(parent)
        }
        Log.e(TAG, "Unknown Word type")
        return TrainingWordRecyclerViewHolder(parent)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val word = words[position]
        (holder as WordsViewHolder).bind(word)
        holder.itemView.isSelected = word.isSelected
        holder.itemView.setOnClickListener { view ->
            val newSelection = !word.isSelected
            word.isSelected = newSelection
            view.isSelected = newSelection
            updateSelection(if (newSelection) selectedItemsCount + 1 else selectedItemsCount - 1)
        }
    }

    private fun updateSelection(newSelectedItemsCount: Int) {
        selectedItemsCount = newSelectedItemsCount

        if (ObjectUtils.isNotNull(TAG, "updateSelection", listener)) {
            if (selectedItemsCount > 1) {
                listener.updateMultipleSelection()
            } else if (selectedItemsCount == 1) {
                listener.updateSingleSelection()
            } else {
                listener.updateNoSelection()
            }
        }
    }

    override fun getItemViewType(position: Int): Int {
        val (_, _, type) = words[position]
        return type
    }

    override fun getItemCount(): Int {
        return words.size
    }

    fun setWords(words: List<WordViewModel>) {
        this.words = words as MutableList<WordViewModel>
        notifyDataSetChanged()
    }

    fun addWord(word: WordViewModel) {
        words.add(word)
        notifyDataSetChanged()
    }

    fun clearSelection() {
        updateSelection(0)
    }

    val selectedWord: WordViewModel
        get() {
            val selectedIndex = firstSelectedIndex
            return if (selectedIndex >= 0)
                words[selectedIndex]
            else
                WordViewModel("", "", Constants.TYPE_TRAINING, false)
        }

    val selectedWords: List<WordViewModel>
        get() {
            return words.filter { it.isSelected }
        }

    private val firstSelectedIndex: Int
        get() {
            return words.indices.firstOrNull { words[it].isSelected } ?: -1
        }

    fun editSelectedWord(word: WordViewModel) {
        words[firstSelectedIndex] = word
        notifyDataSetChanged()
    }

}
